### PR TITLE
Upgrade to Apollo Server v2

### PR DESCRIPTION
**Pull request by mxstbr**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api

**TODO**

- [x] Make fetching the API work again
- [x] Make subscriptions work again
- [x] Fix subscriptions not having the user object in ctx because `context` overrides `onConnect`
- [x] Test and fix file uploads in a backwards compatible manner
- [x] Enable Apollo Engine
- ~~Figure out whether we can use Redis for caching~~ (follow up PR)
- [x] Figure out validation rules
- [x] Upgrade to graphql @ 0.13

### CODE REVIEW

**Comment by mxstbr** [comment]
Ugh file uploads don't work anymore, and the new way of doing it is incompatible with the way our clients do it right now I think. 😕

**Comment by brianlovin** [comment]
> Ugh file uploads don't work anymore, and the new way of doing it is incompatible with the way our clients do it right now I think.

I'll take a look :)

**Comment by jaydenseric** [comment]
@mxstbr it seems you were previously using `apollo-upload-server@5`, which is the same version used in `apollo-server@2`. All versions of `apollo-upload-server` after v4 use v2 of the [GraphQL multipart request spec](https://github.com/jaydenseric/graphql-multipart-request-spec), so in thoery your clients should still work.

With `apollo-server@2` you can remove `apollo-upload-server` from your dependencies, and remove the `scalar Upload` definition from your schema as it takes care of all that itself.

Keep in mind that `apollo-server@2` uses an outdated version of `apollo-upload-server` that has critical issues [fixed in later versions](https://github.com/jaydenseric/apollo-upload-server/blob/master/changelog.md#700). They don't want to upgrade because the new versions lift support from Node.js v6+ to v8.5+, see https://github.com/jaydenseric/apollo-upload-server/issues/109.

**Comment by mxstbr** [comment]
> and remove the scalar Upload definition from your schema as it takes care of all that itself.

The issue is that if I replace the custom scalar that all our client-side queries will break since the types change, right?

```GraphQL
# This breaks since the scalar is now called something else
# so all old clients cannot upload messages anymore
mutation addMessage($file: Upload!) { }
```

**Comment by jaydenseric** [comment]
@mxstbr the scalar is still called `Upload`, you just don't need to set it up manually for `apollo-server@2` is all. It can be used the same as before.

**Comment by mxstbr** [comment]
Sorry, we called it `File` instead of `Upload`... Can we still use the custom scalar manually and then just use that?

**Comment by jaydenseric** [comment]
You're in a pickle. You might be able to experiment and find a way to get `File` working so your API is backwards compatible for clients, but I would suggest migrating to `Upload` as it's pretty standard and has better semantics.

Technically the `Upload` scalar represents the process of uploading a file, not the file itself. For example, the `encoding` property of the resolved promise represents the encoding for the transfer stream, not the encoding of the actual file contents (once it's been fully received). An upload is not guaranteed to result in a file; if the connection is aborted early the upload promise rejects and a readable stream is not resolved.

**Comment by mxstbr** [comment]
> I would suggest migrating to Upload as it's pretty standard and has better semantics.

I want to, but it can't be at the expense of all users on old clients being broken :confused:

**Comment by mxstbr** [comment]
Realized we weren't even using the `File` type anywhere anymore, we are using `Upload` across the board anyway! :tada: File uploads work perfectly now!!

**Comment by jaydenseric** [comment]
> File uploads work perfectly

Well, apollo-server uses an Apollo fork of graphql-upload at v5, which no one should still be using. By not using v8 you miss a lot of critical updates… https://github.com/jaydenseric/graphql-upload/blob/master/changelog.md

Glad you didn't have to rename the `Upload` scalar after all though.

**Comment by mxstbr** [comment]
Oh shit. How can we use v8 instead then? Disable the ApolloServer integration and switch back to the custom `graphqlUploadExpress` middleware?

**Comment by jaydenseric** [comment]
@mxstbr I just DM'd you on spectrum.chat :)

**Review by brianlovin — DISMISSED** [review]
Overall this lgtm @mxstbr - would be awesome if you could stage this rollout while you're around and available to revert :)

File: api/models/usersChannels.js
Comment (brianlovin) [inline api/models/usersChannels.js:640]: Why is this first argument optional?

Comment (mxstbr) [reply api/models/usersChannels.js:640]: Because the user might be logged out, at which point this will just return the default permissions :+1:

**Comment by mxstbr** [comment]
Seems like some stuff is still not working here, will dig in tomorrow!

**Comment by mxstbr** [comment]
This currently blocked by this issue: https://github.com/apollographql/apollo-server/issues/1627

**EDIT**: This turns out to be actually caused by graphql-cost-analysis, ref pa-bru/graphql-cost-analysis#12. commit 9b1f980 uses the workaround from that issue to resolve this for now so we can ship this upgrade.

**Comment by mxstbr** [comment]
Yay it passed! :tada: @brianlovin this is ready for a review, as well as #3957 which builds upon this upgrade.

**Review by brianlovin — APPROVED** [review]
Amazing work man, thanks for pushing through on this! Local testing looks great, connections are good. I'll merge this into alpha, but it'd be good to keep this on alpha.spectrum.chat for a little while to ensure things are smooth live.

**Pull request merged**
